daemon/exec_linux.go

```diff
@@ -41,6 +41,7 @@ func getUserFromContainerd(ctx context.Context, containerdCli *containerd.Client
 		coci.WithUser(ec.User),
 		withResetAdditionalGIDs(),
 		coci.WithAdditionalGIDs(ec.User),
+		coci.WithAppendAdditionalGroups(ec.Container.HostConfig.GroupAdd...),
```

Comment:
This function (and also other `WithUser`/`With*GID`) mounts the container snapshot in a read-only mode:
https://github.com/moby/moby/blob/a66ddb7c59781ab4b7d8b931658fa315157bc29b/vendor/github.com/containerd/containerd/oci/spec_opts.go#L750-L754

which is undefined behavior for overlayfs when a directory is also mounted as a work-dir of other mount.

And this condition is true in this case, because for `exec`, the container will already be running and so the container snapshot (with container working layer) is already mounted.

This is not a new problem though, because this is already happening in the other `WithAdditionalGIDs`.

Reply:
We (@laurazard did) fixed this in containerd https://github.com/containerd/containerd/pull/8299

Reply:
Ah, I missed that this removes the work-dir completely, instead of just moving to the lower-dir.
That solves the issue I think, although that probably means that changes to `/etc/passwd` during the container lifetime won't be picked up by this code? It's probably not a big problem though.

Reply:
Why wouldn't they be picked up?

Reply:
Ah, nevermind, I'm talking nonsense, the upperdir is still in the lower-dir.

In that case, I think the readonly mount will still cause undefined behavior because we mount the container rw-layer (which is an upper-dir of the other mount) as lower-dir.

BUT! It's actually not going to happen there, because we don't even hit this mount code.
We don't even set the `Snapshotter` and `SnapshotterKey` for exec so it will end up taking this branch:
https://github.com/moby/moby/blob/a66ddb7c59781ab4b7d8b931658fa315157bc29b/vendor/github.com/containerd/containerd/oci/spec_opts.go#L732-L736

and will just use the `s.Root.Path` directly.

So, it's fine in this case, but we must remember about it if at some point in future we set `Snapshotter` and `SnapshotterKey` in exec config.

Reply:
> So, it's fine in this case, but we must remember about it if at some point in future we set Snapshotter and SnapshotterKey in exec config.

@vvoland can you create a ticket for this if changes are needed somewhere? (this repo and/or containerd)?

Reply:
Yeah, we will always only set the root path since we are the ones doing the mounting (so that we can reference count things)

But we did set the `Snapshotter` in the past https://github.com/moby/moby/pull/45698/files#diff-034d55db8b341bd127a5394a84226b04c91d54669dd856051ceb0bcbbb50b380L182